Bitcoin Core 0.23 installer incompatible with Windows mandatory ASLR

<!-- Describe the issue -->

Try downloading the current version 0.23 of Bitcoin Core for Windows from the official site : the installer (EXE) is corrupted (won't run, corrupted EXE format)

Only the ZIP file is correct.

Workaround: you can use the 0.22 installer, then download the 0.23 zip and extract the executables from there to overwrite the existing EXE files (qt.exe in the main Program Files\Bitcoin folder, other executables in the ZIP go to the "daemon" subfolder)

So the EXE package was incorrectly built, or was incorrectly transfered/truncated on the download site on bitcoin.org.

I also found that the EXE file contains MORE relocation entries than those indicated in the PE EXE header for the ".text" section.

This is confirmed with "DUMPBIN.EXE" (from MS Visual Studio 2022), which signals it in the ".pdata" section. I get this message (in French because VS Studio 2022 is installed and running on the French system locale) when dumping the relocation entries:

> LINK : fatal error LNK1337: le binaire contient plus de réadressages de base que ce que son répertoire de réadressages de base indique

This does not occur with the v0.22 installer.

So this is a bug in the linker used with your mingw suite (the EXE contains signatures from "GCC: (GNU) 10.3.0".
I think this is a bug of its PE EXE linker.

Just consider using DUMPBIN (the free tool provided by Visual Studio 2022).
- Try running: DUMPBIN /ALL 'bitcoin-setup.exe' >log
- look at the "LINK : fatal error LNK1337" occuring while dumping the ".pdata" section.
- You can detect the GCC version used by looking at the end of the dumped ".text" section in the log file. It is different even if you used the same version of NSIS.
- You can detect the linker version in the headers at top of file. It is different even if you used the same version of NSIS.

There may be equivalent dump tools for the GCC suite (but they make different assumptions, possibly based on Linux capabilities) to inspect the relocation entries, and see by yourself that their effective number don't match what is catalogued in the PE headers. Who's wrong? GCC generating fake/duplicate relocation entries, or the linker not properly updating the EXE header when it terminates its job, or NSIS using GCC with incorrect options?

--- a/contrib/guix/manifest.scm
+++ b/contrib/guix/manifest.scm
@@ -194,7 +194,8 @@ chain for " target " development."))
 
 (define (make-nsis-for-gcc-10 base-nsis)
   (package-with-extra-patches base-nsis
-    (search-our-patches "nsis-gcc-10-memmove.patch")))
+    (search-our-patches "nsis-gcc-10-memmove.patch"
+                        "nsis-disable-installer-reloc.patch")))
 
 (define (fix-ppc64-nx-default lief)
   (package-with-extra-patches lief

--- a/contrib/guix/patches/nsis-disable-installer-reloc.patch
+++ b/contrib/guix/patches/nsis-disable-installer-reloc.patch
@@ -1,0 +1,30 @@
+Patch NSIS so that it's installer stubs, produced at NSIS build time,
+do not contain .reloc sections, which will exist by default when using
+binutils/ld 2.36+.
+
+This ultimately fixes an issue when running the installer with the
+"Force randomization for images (Mandatory ASLR)" setting active.
+
+This patch has not yet been sent upstream, because it's not clear if this
+is the best fix, for the underlying issue, which seems to be that makensis
+doesn't account for .reloc sections when it builds installers.
+
+The existence of a reloc section shouldn't be a problem, and, if anything,
+is actually a requirement for working ASLR. All other Windows binaries we
+produce contain them, and function correctly when under the same
+"Force randomization for images (Mandatory ASLR)" setting.
+
+See:
+https://github.com/bitcoin/bitcoin/issues/25726
+https://sourceforge.net/p/nsis/bugs/1131/
+
+--- a/SCons/Config/gnu
++++ b/SCons/Config/gnu
+@@ -102,6 +102,7 @@ stub_env.Append(LINKFLAGS = ['-mwindows'])          # build windows executables
+ stub_env.Append(LINKFLAGS = ['$NODEFLIBS_FLAG'])    # no standard libraries
+ stub_env.Append(LINKFLAGS = ['$ALIGN_FLAG'])        # 512 bytes align
+ stub_env.Append(LINKFLAGS = ['$MAP_FLAG'])          # generate map file
++stub_env.Append(LINKFLAGS = ['-Wl,--disable-reloc-section'])
+ 
+ conf = FlagsConfigure(stub_env)
+ conf.CheckCompileFlag('-fno-tree-loop-distribute-patterns')  # GCC 10: Don't generate msvcrt!memmove calls (bug #1248)

